csharp_dotnetcore/9.Message-Routing/Startup.cs

```diff
@@ -0,0 +1,154 @@
+        public void ConfigureServices(IServiceCollection services)
+        {
+            // Loads .bot configuration file and adds a singleton that your Bot can access through dependency injection.
+            var botConfig = BotConfiguration.LoadAsync(@".\MessageRouting.bot").GetAwaiter().GetResult();
+            services.AddSingleton(sp => botConfig);
+
+            // Initializes your bot service clients and adds a singleton that your Bot can access through dependency injection.
+            var connectedServices = InitBotServices(botConfig);
+            services.AddSingleton(sp => connectedServices);
+
+            // Initializes Bot Conversation and User State and adds a singleton that your Bot can access through dependency injection.
+            services.AddSingleton(sp =>
+            {
+                var options = sp.GetRequiredService<IOptions<BotFrameworkOptions>>().Value;
+                var conversationState = options.State.OfType<ConversationState>().FirstOrDefault();
+                var userState = options.State.OfType<UserState>().FirstOrDefault();
+
+                var accessors = new MessageRoutingBotAccessors
+                {
+                    ConversationDialogState = conversationState.CreateProperty<DialogState>("DialogState"),
+                };
+
+                return accessors;
+            });
+
+            services.AddBot<MessageRoutingBot>(options =>
+            {
+                options.CredentialProvider = new ConfigurationCredentialProvider(this.Configuration);
+
+                // Catches any errors that occur during a conversation turn and logs them to AppInsights.
+                options.OnTurnError = async (context, exception) =>
+                {
+                    await context.SendActivityAsync("Sorry, it looks like something went wrong.");
+                };
+
+                // Memory Storage is for local bot debugging only. When the bot
+                // is restarted, everything stored in memory will be gone.
+                IStorage dataStore = new MemoryStorage();
+
+                // For production bots use the Azure Blob or
+                // Azure CosmosDB storage providers, as seen below. To the Azure
+                // based storage providers, add the Microsoft.Bot.Builder.Azure
+                // Nuget package to your solution. That package is found at:
+                // https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/
+                // Uncomment this line to use Azure Blob Storage
+                // IStorage dataStore = new Microsoft.Bot.Builder.Azure.AzureBlobStorage("AzureBlobConnectionString", "containerName");
+                // Create and add conversation state.
+                var convoState = new ConversationState(dataStore);
+                options.State.Add(convoState);
+                options.Middleware.Add(new BotStateSet(options.State.ToArray()));
+            });
```

Comment:
Remove BotStateSet in alignment with our other samples. Just call "Save" in the main onTurn method. Makes things much simpler acorss the board, and developers can handle save errors if they happen.

Reply:
@cleemullins do you have an example of this? Is this save the one you mean?
https://github.com/Microsoft/BotBuilder-Samples/blob/v4/csharp_dotnetcore/5.MultiTurn-Prompts-Bot/MultiTurnPromptsBot.cs#L96

Reply:
Yes, that one is fine. Should take about 2 mintues to make this change... 

---
In reply to: [216499197](https://github.com/Microsoft/BotBuilder-Samples/pull/349#discussion_r216499197) [](ancestors = 216499197)